trending_data/src/main/java/com/google/sps/servlets/TweetsGetterServlet.java

```diff
@@ -0,0 +1,46 @@
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import twitter4j.Query;
+import twitter4j.QueryResult;
+import twitter4j.Status;
+import twitter4j.Twitter;
+import twitter4j.TwitterException;
+import twitter4j.TwitterFactory;
+
+@WebServlet("/get-tweets")
+public class TweetsGetterServlet extends HttpServlet {
+  // The maximum number of tweets returned
+  static final int NUMBER_OF_TWEETS = 5;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json; charset=UTF-8");
+    String query = request.getParameter("query");
+    try {
+      List<String> tweetIds = getTweetIds(query);
+      response.getWriter().println(new Gson().toJson(tweetIds));
+    } catch (TwitterException twException) {
+      response.getWriter().println(new Gson().toJson(Collections.emptyList()));
+    }
+  }
+  /**
+   * Returns a list of strings of tweet ids corresponding to the given queries. The size of the list
+   * <= NUMBER_OF_TWEETS.
+   */
+  private List<String> getTweetIds(String query) throws TwitterException {
+    Twitter twitter = new TwitterFactory().getInstance();
+    QueryResult qr = twitter.search().search(new Query(query).count(NUMBER_OF_TWEETS));
```

Comment:
Does this result return only the ids? Or also the tweet content? 

Right now, only the ids are sent to the frontend. So I believe the twitter element on the frontend reissues a request to twitter to then fetch the tweets. That would explain the delay you mentioned. 

If so, is there a way to instead send the tweet content/metadata to the frontend so it doesn't need to be re-fetched?

Reply:
The library that I use to get the "tweet cards" and put them on the page needs the id only. I don't know another way to build a tweet card but I can look into it.